docs/RFCS/20190619_constraint_conformance_report.md

```diff
@@ -0,0 +1,319 @@
+their configured “replication factor”).
+  E.g.
+
+    ```
+    table foo, replication factor, 1/1000 ranges have only 2 replicas, 2/1000 ranges
+    have 4 replicas (out of these 2, 1/2 ranges are also under-diversified having 2
+    replicas in dc=west)
+    ```
+1. How many ranges/how much data violates each constraint.
+  E.g. 
+
+    ```
+    table foo, constraint "2 replicas in dc=east, 1 replica in dc=west", 30/1000
+    ranges (1GB/30GB of data) are in violation (not enough replicas in dc=east)
+    
+    table baz partition "north_america", constraint "region=us", 20/1000 ranges are
+    in violation (replicas outside region=us)
+    ````
+1. How much data violates the leaseholder preference.
+1. How much data is under-diversified.
+  Data is considered to be under-diversified if the allocator would like to
+  move a replica away to improve it. If there's no locality to move a replica
+  to, then the range is not under-diversified. So, for example, if there is a
+  single region (and so all the replicas are in that region), there's no
+  under-diversification (but there will be the moment a second region is
+  added). Similarly, say there are 3 regions each with 3 AZs. Absent any
+  constraints, there'll be a replica in each region, in a random AZ. This again
+  is fine. On the other hand, lack storage capacity in a locality is no excuse
+  for under-diversification - i.e. we'll report a range as under-diversified
+  even if no replica can be moved at the moment because of space constraints. 
+  Note that when the under-diversification is due to over-replication (i.e. the
+  only violation is that two replicas are in the same locality), that’d be
+  accounted for in the over-replication stats, without also being counted here.
+
+
+Besides raw numbers, this report will also include velocity information (the
+`change_rate` col): how did each number change in the last *x* minutes. This can
+be `NULL` if there's no previous data on a particular violation.
+
+## AdminUI conformance report
+
+We’ll add a page in the AdminUI showing the contents of the virtual table. In
+the AdminUI we can do fancier things like allowing one to drill down into a zone
+- expand a zone into its constituent tables.
+
+## New jobs
+
+We’re also going to have some data-moving operations roughly represented as
+jobs: repartitioning and altering zone config properties. The idea here is to be
+able to track the progress of an operation that was explicitly triggered through
+a SQL command as opposed to observing the state of the cluster as a whole.
+For example, when a table is repartitioned, we’ll look at how much data of that
+table is now placed incorrectly (i.e. it violates a constraint) and we’ll
+consider the job complete once all the table’s data is conformant with all the
+constraints. Note that the initial partitioning of a table does not create any
+data movement (as all the partitions inherit the same zone config) and so we
+won’t create a job for it.
```

Comment:
do we need to mention that we will mark it complete the first time the jobs criterion are satisfied? in the scenario in which this data can be in flux for any number of reasons

Reply:
i see now its mentioned below but maybe it should be both places?

Reply:
ack